File: .skills/openclaw-skills/skills/zaohon/dingtalk-bot-publish/utils/http.ts
// HTTP请求工具模块
// 用于统一处理钉钉API的HTTP请求

import * as https from 'https';

/**
 * 钉钉API请求选项接口
 */
interface DingTalkRequestOptions {
  method: string;
  path: string;
  accessToken: string;
  body?: any;
  headers?: Record<string, string>;
}

/**
 * 执行钉钉API请求
 * @param options 请求选项
 * @returns Promise<any> API响应数据
 */
export async function dingtalkRequest(options: DingTalkRequestOptions): Promise<any> {
  const { method, path, accessToken, body, headers = {} } = options;

  return new Promise((resolve, reject) => {
    const requestOptions = {
      hostname: 'oapi.dingtalk.com',
      path: `${path}?access_token=${accessToken}`,
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
    };

    const req = https.request(requestOptions, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          
          // 检查钉钉API错误码
          if (parsed.errcode !== undefined && parsed.errcode !== 0) {
            reject({
              code: parsed.errcode,
              message: parsed.errmsg || 'Unknown error',
              requestId: parsed.requestId || null,
            });
          } else if (res.statusCode && res.statusCode >= 400) {
            reject({
              code: res.statusCode,
              message: `HTTP ${res.statusCode} error`,
              data: parsed,
            });
          } else {
            resolve(parsed);
          }
        } catch (error) {
          reject(new Error(`Invalid JSON response: ${data}`));
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    if (body) {
      req.write(JSON.stringify(body));
    }
    req.end();
  });
}

/**
 * 构建标准的成功响应
 * @param data 响应数据
 * @returns 标准化成功响应对象
 */
export function createSuccessResponse(data: any): any {
  return {
    success: true,
    ...data,
  };
}

/**
 * 构建标准的错误响应
 * @param error 错误信息
 * @param userId 可选的用户ID
 * @returns 标准化错误响应对象
 */
export function createErrorResponse(error: any, userId?: string): any {
  return {
    success: false,
    ...(userId ? { userId } : {}),
    error: {
      code: error.code || 'UNKNOWN_ERROR',
      message: error.message || error.msg || JSON.stringify(error),
      description: error.description || null,
      requestId: error.requestId || null,
    },
  };
}

File: .skills/openclaw-skills/skills/vihangd/scry/scripts/lib/http.py
"""HTTP utilities for SCRY skill (stdlib only)."""

import json
import os
import sys
import time
import urllib.error
import urllib.request
from typing import Any, Dict, Optional
from urllib.parse import urlencode

DEFAULT_TIMEOUT = 30
DEBUG = os.environ.get("SCRY_DEBUG", "").lower() in ("1", "true", "yes")


def log(msg: str):
    if DEBUG:
        sys.stderr.write(f"[scry] {msg}\n")
        sys.stderr.flush()


MAX_RETRIES = 5
RETRY_DELAY = 2.0
USER_AGENT = "scry-skill/1.0 (Claude Code Skill)"


class HTTPError(Exception):
    def __init__(self, message: str, status_code: Optional[int] = None, body: Optional[str] = None):
        super().__init__(message)
        self.status_code = status_code
        self.body = body


def request(
    method: str,
    url: str,
    headers: Optional[Dict[str, str]] = None,
    json_data: Optional[Dict[str, Any]] = None,
    timeout: int = DEFAULT_TIMEOUT,
    retries: int = MAX_RETRIES,
    raw: bool = False,
) -> Any:
    headers = headers or {}
    headers.setdefault("User-Agent", USER_AGENT)

    data = None
    if json_data is not None:
        data = json.dumps(json_data).encode("utf-8")
        headers.setdefault("Content-Type", "application/json")

    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    log(f"{method} {url}")

    last_error = None
    for attempt in range(retries):
        try:
            with urllib.request.urlopen(req, timeout=timeout) as response:
                body = response.read().decode("utf-8")
                log(f"Response: {response.status} ({len(body)} bytes)")
                if raw:
                    return body
                return json.loads(body) if body else {}
        except urllib.error.HTTPError as e:
            body = None
            try:
                body = e.read().decode("utf-8")
            except Exception:
                pass
            log(f"HTTP Error {e.code}: {e.reason}")
            last_error = HTTPError(f"HTTP {e.code}: {e.reason}", e.code, body)
            if 400 <= e.code < 500 and e.code != 429:
                raise last_error
            if attempt < retries - 1:
                if e.code == 429:
                    retry_after = e.headers.get("Retry-After") if hasattr(e, "headers") else None
                    if retry_after:
                        try:
                            delay = float(retry_after)
                        except ValueError:
                            delay = RETRY_DELAY * (2 ** attempt) + 1
                    else:
                        delay = RETRY_DELAY * (2 ** attempt) + 1
                    log(f"Rate limited. Waiting {delay:.1f}s (retry {attempt + 2}/{retries})")
                else:
                    delay = RETRY_DELAY * (2 ** attempt)
                time.sleep(delay)
        except urllib.error.URLError as e:
            log(f"URL Error: {e.reason}")
            last_error = HTTPError(f"URL Error: {e.reason}")
            if attempt < retries - 1:
                time.sleep(RETRY_DELAY * (attempt + 1))
        except json.JSONDecodeError as e:
            last_error = HTTPError(f"Invalid JSON response: {e}")
            raise last_error
        except (OSError, TimeoutError, ConnectionResetError) as e:
            log(f"Connection error: {type(e).__name__}: {e}")
            last_error = HTTPError(f"Connection error: {type(e).__name__}: {e}")
            if attempt < retries - 1:
                time.sleep(RETRY_DELAY * (attempt + 1))

    if last_error:
        raise last_error
    raise HTTPError("Request failed with no error details")


def get(url: str, headers: Optional[Dict[str, str]] = None, **kwargs) -> Any:
    return request("GET", url, headers=headers, **kwargs)


def post(url: str, json_data: Dict[str, Any], headers: Optional[Dict[str, str]] = None, **kwargs) -> Any:
    return request("POST", url, headers=headers, json_data=json_data, **kwargs)


def post_raw(url: str, json_data: Dict[str, Any], headers: Optional[Dict[str, str]] = None, **kwargs) -> str:
    return request("POST", url, headers=headers, json_data=json_data, raw=True, **kwargs)


def get_reddit_json(path: str, timeout: int = DEFAULT_TIMEOUT, retries: int = MAX_RETRIES) -> Dict[str, Any]:
    if not path.startswith("/"):
        path = "/" + path
    path = path.rstrip("/")
    if not path.endswith(".json"):
        path = path + ".json"
    url = f"https://www.reddit.com{path}?raw_json=1"
    headers = {"User-Agent": USER_AGENT, "Accept": "application/json"}
    return get(url, headers=headers, timeout=timeout, retries=retries)


def fetch_rss(url: str, timeout: int = DEFAULT_TIMEOUT) -> str:
    """Fetch RSS/Atom feed and return raw XML text."""
    headers = {"User-Agent": USER_AGENT, "Accept": "application/rss+xml, application/atom+xml, application/xml, text/xml"}
    return request("GET", url, headers=headers, timeout=timeout, retries=2, raw=True)
